src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRetryHandler.cs

```diff
@@ -197,20 +203,50 @@ internal HttpRetryHandler(IEnvironmentVariableReader environmentVariableReader)
                             response.Content = newContent;
                         }
 
-                        log.LogInformation("  " + string.Format(
-                            CultureInfo.InvariantCulture,
-                            Strings.Http_ResponseLog,
-                            response.StatusCode,
-                            requestUri,
-                            bodyStopwatch.ElapsedMilliseconds));
+                        retryAfter = GetRetryAfter(response.Headers.RetryAfter);
+                        if (retryAfter != null)
+                        {
+                            log.LogInformation("  " + string.Format(
+                                CultureInfo.InvariantCulture,
+                                Strings.Http_ResponseLogWithRetryAfter,
+                                response.StatusCode,
+                                requestUri,
+                                bodyStopwatch.ElapsedMilliseconds,
+                                retryAfter.Value.TotalSeconds));
+
+                            if (retryAfter.Value.TotalMilliseconds < 0)
+                            {
+                                retryAfter = null;
+                            }
+                            else if (retryAfter.Value > _enhancedHttpRetryHelper.MaxRetryAfterDelay)
+                            {
+                                retryAfter = _enhancedHttpRetryHelper.MaxRetryAfterDelay;
+                            }
+                        }
+                        else
+                        {
+                            log.LogInformation("  " + string.Format(
+                                CultureInfo.InvariantCulture,
+                                Strings.Http_ResponseLog,
+                                response.StatusCode,
+                                requestUri,
+                                bodyStopwatch.ElapsedMilliseconds));
+                        }
```

Comment:
I think the entire logic can be extracted into class that inherits `DelegatingHandler` and inject the new handler into the HTTP handler pipeline. I have documented all the HTTP handlers involved in processing HTTP requests in [this](https://github.com/NuGet/Client.Engineering/blob/main/designs/httpstohttp-redirection-netfx.md#option-2---create-a-new-http-handler-to-respond-to-redirects) doc. I am sure it may add additional complexity, but it keeps the responsibility of new class limited to the way how we react to 429/503 responses and don't have to modify exponential enhanced retry helper logic when we have to change how we respond to 429/503 responses.

For example, [dotnet/arcade-services/RetryAfterHandler.cs](https://github.com/dotnet/arcade-services/blob/main/src/Shared/Microsoft.DotNet.Services.Utility/RetryAfterHandler.cs)

Reply:
This is already a retry handler? Why should we have 2 separate handler? I understand that while related, they're still slightly different, but I'd hate to just keep adding abstractions on top of abstractions. This code would only be used in 1 place anyways, so I don't see the point of another abstraction.

Reply:
I am okay either way. My understanding is that whenever we trigger a HTTP request it flows through the following handlers called HTTP processing pipeline. My suggestion is, if we add a new `RetryAfterHandler` in between `ProxyAuthenticationHandler` and `ServerWarningLogHandler` and if the server responds with `429/503` then we don't have execute all the handlers and then issue a retry request from `HttpRetryHandler` level.

```mermaid
sequenceDiagram
    HttpRetryHandler->>+HttpSourceAuthenticationHandler:Send request
    HttpSourceAuthenticationHandler->>+StsAuthenticationHandler:Adds source credentials
    StsAuthenticationHandler->>+ProxyAuthenticationHandler:Adds Sts token
    ProxyAuthenticationHandler->>+ServerWarningLogHandler:Adds Proxy details
    ServerWarningLogHandler->>+HttpClientHandler:Send request to .NET
    HttpClientHandler-->>-ServerWarningLogHandler:Returns response (auto redirects)
    ServerWarningLogHandler-->>ProxyAuthenticationHandler:Returns response (log warnings)
    ProxyAuthenticationHandler-->>StsAuthenticationHandler:Returns response (retries)
    StsAuthenticationHandler-->>HttpSourceAuthenticationHandler:Returns response (retries)
    HttpSourceAuthenticationHandler-->>HttpRetryHandler:Returns response (retries)
```

Reply:
Moving retries into HttpClient's pipeline has some practical constraints like passing the `ILogger`, so that retries can be logged.

Also, it doesn't solve the problem that some resources (like [`ServiceIndexResourceV3Provider` implementing its own retries](https://github.com/NuGet/NuGet.Client/blob/a487c8336b29a0e53dc46f4aab424fdd5de50e72/src/NuGet.Core/NuGet.Protocol/Providers/ServiceIndexResourceV3Provider.cs#L131). At least with the current implementation, these resources can send `HttpSource` and `HttpRetryHandler` information via parameters about how many retries to make (to avoid retries being multiplicative).

Finally, consider that files are downloaded with `HttpCompletionOptions.ResponseHeadersRead`. This means that the HttpClient (and any handlers) return when the headers are received, but the content stream hasn't yet complete. If the download fails mid-stream, then the code that calls `HttpClient.SendAsync` still needs its own retry logic, and now we're got 2 retry handlers, and they need to somehow synchronize the number of retries that's been attempted, assuming we want to honor MaxRetries/attempt number correctly (otherwise exponention backoff can't work). This also includes resetting the destination file stream, which `HttpClient` and its internal handlers, never see.

On the other hand, handling retries outside of HttpClient (in the code that calls HttpClient.SendAsync), then we can have a single implementation that handles retries both un-successful HTTP status codes, as well as downloads failing mid-stream. It also avoids all the other problems listed above.